nova/scheduler/filters/shard_filter.py

```diff
@@ -114,11 +125,106 @@ def _get_shards(self, project_id):
 
         return self._PROJECT_SHARD_CACHE.get(project_id)
 
-    def host_passes(self, host_state, spec_obj):
+    def _get_k8s_shard(self, spec_obj):
+        """Returns the dominant shard of a K8S cluster.
+
+        Returns None if the request is not for an instance that's part of
+        a K8S cluster, or if this is the first instance of a new cluster.
+        """
+        if (spec_obj.flavor.name.startswith(HANA_PREFIX) or
+                utils.request_is_resize(spec_obj)):
+            return None
+        elevated = nova_context.get_admin_context()
+        build_request = None
+        instance = None
+
+        def _get_tags():
+            return build_request.tags if build_request \
+                else instance.tags
+
+        def _get_metadata():
+            return build_request.instance.metadata if build_request \
+                else instance.metadata
+
+        check_type = spec_obj.get_scheduler_hint('_nova_check_type')
+        if not check_type:
+            build_request = BuildRequest.get_by_instance_uuid(
+                elevated, spec_obj.instance_uuid)
+        else:
+            instance = Instance.get_by_uuid(
+                elevated, spec_obj.instance_uuid,
+                expected_attrs=['tags', 'metadata'])
+
+        kks_tag = next((t.tag for t in _get_tags()
+                        if t.tag.startswith(KKS_PREFIX)), None)
+        gardener_meta = None
+        if not kks_tag:
+            gardener_meta = \
+                {k: v for k, v in _get_metadata().items()
+                 if k.startswith(GARDENER_PREFIX)}
+
+        if not kks_tag and not gardener_meta:
+            return None
+
+        q_filters = {'hv_type': VMWARE_HV_TYPE}
+        if spec_obj.availability_zone:
+            q_filters['availability_zone'] = spec_obj.availability_zone
+
+        results = None
+        if kks_tag:
+            results = nova_context.scatter_gather_skip_cell0(
+                elevated, main_db_api.get_k8s_hosts_by_instances_tag,
+                kks_tag, filters=q_filters)
+        elif gardener_meta:
+            (meta_key, meta_value) = next(iter(gardener_meta.items()))
+            results = nova_context.scatter_gather_skip_cell0(
+                elevated, main_db_api.get_k8s_hosts_by_instances_metadata,
+                meta_key, meta_value, filters=q_filters)
+
+        if not results:
+            return None
+
+        # hosts with count of instances from this K8S cluster
+        # {host: <count>}
+        k8s_hosts = {}
+
+        for cell_result in results.values():
+            if not nova_context.is_cell_failure_sentinel(cell_result):
+                cell_hosts = dict(cell_result)
+                k8s_hosts = {
+                    h: k8s_hosts.get(h, 0) + cell_hosts.get(h, 0)
+                    for h in set(cell_hosts) | set(k8s_hosts)
+                }
```

Comment:
Suggestion:
With a `defaultdict`, this could look like:
```python
k8s_hosts = defaultdict(lambda: 0)
for cell_result in results.values():
    if nova_context.is_cell_failure_sentinel(cell_result):
        continue
    for h, c in cell_hosts.items():
        k8s_hosts[h] += c
```

I think this would be easier readable.

Additionally: Is there a way for a host to turn up in multiple cells? If not, shouldn't we be able to use `k8s_hosts.update(cell_hosts)` after checking that it's not a failure?

Additionally: Should we really schedule if we have a failure? The majority of the cluster might reside in the failed shard.